packages/block-editor/src/components/link-control/index.js

```diff
@@ -156,6 +186,19 @@ function LinkControl( {
 		return couldBeURL && ! args.isInitialSuggestions ? results[ 0 ].concat( results[ 1 ] ) : results[ 0 ];
 	};
 
+	/**
+	 * Cancels editing state and marks that focus may need to be restored after
+	 * the next render, if focus was within the wrapper when editing finished.
+	 */
+	function stopEditing() {
+		isEndingEditWithFocus.current = (
+			wrapperNode.current &&
+			wrapperNode.current.contains( document.activeElement )
```

Comment:
I guess as of #19831 I could simplify this using optional chaining:

```suggestion
			wrapperNode.current?.contains( document.activeElement )
```

What's interesting to me is looking at the other, negated instance of this condition, I sense there's going to be a very common pitfall associated with optional chaining:

https://github.com/WordPress/gutenberg/blob/5b72b3a9b38da973cfd4bedd71a3210ddb607d45/packages/block-editor/src/components/link-control/index.js#L112-L113

If someone treats optional chaining as a way to _get to the value by any means necessary_, it might be easy to overlook that negating would produce `true` if the optional chaining aborts early:

```js
const wrapperNode = { current: undefined };
! wrapperNode.current?.contains( document.activeElement );
// true
```

😯 It's definitely not the expected result.

It seems like in the same category of issues we restrict with this ESLint rule:

https://eslint.org/docs/rules/no-unsafe-negation

I could imagine some improvement to that rule which forbids negation of optional chains.

cc @gziolo @sainthkh (more an FYI, as it was interesting to me)

Reply:
Another interesting pitfall: If you're expecting a boolean out of a value or function result at the end of an optional chain, you might be surprised to find that a non-boolean can be assigned if the chain aborts early, since it will evaluate to the value at the point at which it aborted.

```js
const wrapperNode = { current: undefined };
const hasFocus = wrapperNode.current?.contains( document.activeElement );
console.log( typeof hasFocus === 'boolean' );
// `false`
```

This is similar to what can happen when using `&&`, where if the left-hand is _falsy_, the expression doesn't evaluate to an explicit `false`, but rather to the left-hand value.

You may have run into this one before 😄 https://codepen.io/aduth/pen/dWwzrL

It's not often a problem as long as the developer proceeds to use the value in a way which considers it as [_truthy_](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) or [_falsy_](https://developer.mozilla.org/en-US/docs/Glossary/Falsy), but it could be important to consider for satisfying function contracts, where we'd want to be considerate to adhere to the expected types.

Reply:
Very interesting case, thanks for sharing. To make it more interesting, let me share a bit different example which in the different context might also produce an unexpected result:

```js
const wrapperNode = { current: 0 };
wrapperNode.current && ! wrapperNode.current.contains( document.activeElement );
// 0
```

It's always a challenge to make such checks work as intended. The only answer is to learn JavaScript deeply or cover code with tests 😅

Reply:
Yeah, I just now observe that the same `&&` applies to the current proposed implementation, at least if the expectation is that it would always assign a boolean, that's not currently happening (it can assign `undefined`).